### PR TITLE
Add BBEs related to the data.yaml module

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

This BBEs are related to the [data.yaml module](https://github.com/ballerina-platform/ballerina-library/issues/6089) implementation.

### CODE REVIEW

File: examples/anydata-to-yaml-string/anydata_to_yaml_string.bal [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+public function main() returns error? {
+
+    ServerConfig serverConfig = {
+        database: {dbName: "userDB", username: "testUser"},
+        port: 3000,
+        host: "localhost"
+    };
+
+    // Serialize a ballerina value to a YAML value.
```
Comment (MaryamZi) [inline examples/anydata-to-yaml-string/anydata_to_yaml_string.bal:23]:
```suggestion
    // Serialize a Ballerina value to a string in YAML format.
```

File: examples/index.json [hunk 2]
```diff
@@ -4619,6 +4619,34 @@
       }
     ]
   },
+  {
+    "title": "YAML data",
+    "column": 2,
```
Comment (MaryamZi) [inline examples/index.json:4624]: Why column 2 in between column 3 stuff?

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    DatabaseConfig database;
+|};
+
+type DatabaseConfig record {|
+    string dbName;
+    string username;
+|};
+
+
+final string yamlString =
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:17]: Why not use a string template? I would make this a local variable.

Comment (MaryamZi) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:17]: We can add a comment to say this could be something read from a file.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively convert elements and attributes from a YAML source,
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: Line breaks aren't required, right?

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    ServerConfig serverConfig2 = check yaml:parseBytes(yamlByteArr);
+    io:println(serverConfig2);
+
+    stream<byte[], error?> byteBlockStream = new (new ByteBlockGenerator(yamlString));
+    // Convert the YAML byte block stream to a record type.
+    ServerConfig serverConfig3 = check yaml:parseStream(byteBlockStream);
+    io:println(serverConfig3);
+}
+
+// Defines a class called `ByteBlockGenerator`, which implements the `next()` method.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:31]: More helpful to introduce as a stream implementor.

"a class called `ByteBlockGenerator`" is self-explanatory?

File: examples/index.json (hunk 2 above)
Comment (LakshanWeerasinghe) [reply examples/index.json:4624]: This should be 3. Thanks.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal (hunk 3 above)
Comment (LakshanWeerasinghe) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:17]: With string template to add new line characters need to add interpolations.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md (hunk 4 above)
Comment (LakshanWeerasinghe) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: No. Added them to improve readability.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively convert elements and attributes from a YAML source, 
+which can be provided as a `string`, `byte array`, or `byte block stream`, into a Ballerina record. 
+In projection users can drop attributes and elements in arrays when creating the Ballerina value.
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:5]: Will update.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal (hunk 3 above)
Comment (MaryamZi) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:17]: No? Actually becomes a bit easier to read.

```ballerina
final string yamlString = string 
`host: "localhost"
port: 8080
remotePorts: [9000, 9001, 9002, 9003]
protocol: "http"
database:
  dbName: "testdb"
  username: "dbuser"
  password: "dbpassword"
`;
```

Comment (LakshanWeerasinghe) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal]: updated.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md (hunk 4 above)
Comment (MaryamZi) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: Not necessary, could mess up bio rendering.

File: examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags [hunk 7]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates how to convert a subtype of `anydata` value to a YAML format string.
```
Comment (MaryamZi) [inline examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags:1]:
```suggestion
description: This BBE demonstrates how to serialize a value belonging to `anydata` to a string in YAML format.
```

File: examples/anydata-to-yaml-string/anydata_to_yaml_string.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# Anydata to YAML string conversion
```
Comment (MaryamZi) [inline examples/anydata-to-yaml-string/anydata_to_yaml_string.md:1]:
```suggestion
# Serialize a Ballerina value to a string in YAML format
```

File: examples/index.json [hunk 9]
```diff
@@ -4619,6 +4619,34 @@
       }
     ]
   },
+  {
+    "title": "YAML data",
+    "column": 3,
```
Comment (MaryamZi) [inline examples/index.json:4624]: This column may be too long now.

Also see https://github.com/ballerina-platform/ballerina-distribution/issues/5537

Comment (MaryamZi) [reply examples/index.json:4624]: No need to fix others, but can we at least make sure this is added okay?

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    DatabaseConfig database;
+|};
+
+type DatabaseConfig record {|
+    string dbName;
+    string username;
+|};
+
+public function main() returns error? {
+    // Can read from a file as well.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:17]: Change similar to the release note suggestion.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively convert elements and attributes from a YAML source,
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: Multiple APIs?

Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: Can we uniformly use parse instead of convert for parse scenarios?

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively convert elements and attributes from a YAML source, 
+in the form of a string, byte array, or byte block stream, into a Ballerina record. 
+Using projection, users can selectively add attributes to records and limit the sizes of arrays.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:5]: Add an empty line before this.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively convert elements and attributes from a YAML source, 
+in the form of a string, byte array, or byte block stream, into a Ballerina record. 
+Using projection, users can selectively add attributes to records and limit the sizes of arrays. 
+In this example, the `password` attribute is excluded when creating the `DatabaseConfig` record, 
+and only the first two elements from the source are used to create the `remotePorts` array.
+
+
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:9]: Extra new line or do we have two in others?

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    ServerConfig serverConfig2 = check yaml:parseBytes(yamlByteArr);
+    io:println(serverConfig2);
+
+    stream<byte[], error?> byteBlockStream = new (new StreamImplementor(yamlString));
+    // Parse the YAML byte block stream to a record type.
+    ServerConfig serverConfig3 = check yaml:parseStream(byteBlockStream);
+    io:println(serverConfig3);
+}
+
+// Defines a class called `StreamImplementor`, which implements the `next()` method.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:33]: What I meant was to say something along the lines of Define a class that is a stream implementor, ...

File: examples/yaml-to-anydata/yaml_to_anydata.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
+# YAML to anydata conversion
+
+The `data.yaml` library provides multiple APIs to perform the conversion from YAML source, which can be provided as a `string`, `byte array`, or `byte block stream`,
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.md:3]: Fix backticks in `string`, `byte array`, or `byte block stream`,

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md (hunk 11 above)
Comment (LakshanWeerasinghe) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: yes, parseString, parseBytes and parseStream

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md (hunk 13 above)
Comment (LakshanWeerasinghe) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:9]: extra one. Will remove it.

File: examples/index.json (hunk 9 above)
Comment (LakshanWeerasinghe) [reply examples/index.json:4624]: updated the Table to column 2

File: examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags
Comment (MaryamZi) [inline examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags:1]: Why any_data and not anydata?

File: examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags [hunk 16]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates how to serialize a value belonging to `anydata` to a string in YAML format.
+keywords: ballerina, ballerina by example, BBE, yaml, yaml to string, record to yaml
```
Comment (MaryamZi) [inline examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags:2]: yaml to string?

File: examples/anydata-to-yaml-string/anydata_to_yaml_string.md [hunk 17]
```diff
@@ -0,0 +1,9 @@
+# Serialize a Ballerina value to a string in YAML format
+
+The `data.yaml` library provides the `toYamlString` function to convert a value belonging to `anydata` to a string in YAML format.
```
Comment (MaryamZi) [inline examples/anydata-to-yaml-string/anydata_to_yaml_string.md:3]: Please check all usages of convert and change to serialize, parse, etc.

File: examples/index.json (hunk 9 above)
Comment (MaryamZi) [reply examples/index.json:4624]: Missed that this was being added under common libraries. If so, the table change isn't relevant?

Btw we have xmldata and jsondata examples under Language concepts.

File: examples/index.json [hunk 18]
```diff
@@ -4619,6 +4619,34 @@
...
+      },
+      {
+        "name": "YAML to anydata with projection",
+        "url": "yaml-to-anydata-with-projection",
+        "verifyBuild": true,
+        "verifyOutput": true,
+        "isLearnByExample": true
+      },
+      {
+        "name": "Serialize a Ballerina value to a string in YAML format",
```
Comment (MaryamZi) [inline examples/index.json:4642]:
```suggestion
        "name": "Serialize to YAML string",
```

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 19]
```diff
@@ -0,0 +1,14 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively parse elements and attributes from a YAML source,
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:3]: Let's get rid of the unnecessary line breaks.

File: examples/yaml-to-anydata/yaml_to_anydata.md [hunk 20]
```diff
@@ -0,0 +1,9 @@
+# YAML to anydata conversion
+
+The `data.yaml` library provides multiple APIs to parse YAML source, in the form of a string, byte array, or byte block stream, into a subtype of Ballerina `anydata`.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.md:3]:
```suggestion
The `data.yaml` library provides multiple functions to parse YAML source, in the form of a string, byte array, or byte block stream, as a value that belongs to a subtype of `anydata`.
```

File: examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags (hunk 16 above)
Comment (LakshanWeerasinghe) [reply examples/anydata-to-yaml-string/any_data_to_yaml_string.metatags:2]: The serialisation part

File: examples/index.json (hunk 9 above)
Comment (MaryamZi) [reply examples/index.json:4624]: Ping on this.

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md [hunk 21]
```diff
@@ -0,0 +1,11 @@
+# YAML to anydata conversion with projection
+
+The `data.yaml` library provides multiple APIs to selectively parse elements and attributes from a YAML source, in the form of a string, byte array, or byte block stream, into a Ballerina record. Using projection, users can selectively add attributes to records and limit the sizes of arrays.
+
+In this example, the `password` attribute is excluded when creating the `DatabaseConfig` record, and only the first two elements from the source are used to create the `remotePorts` array.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.md:5]: This would be better as or also in the comment in the code.

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 22]
```diff
@@ -0,0 +1,53 @@
+import ballerina/data.yaml;
+import ballerina/io;
+
+type ServerConfig record {|
+    string host;
+    int port;
+    string protocol;
+|};
+
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:9]: Extra new line.

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+    ServerConfig serverConfig2 = check yaml:parseBytes(yamlByteArr);
+    io:println(serverConfig2);
+
+    stream<byte[], error?> byteBlockStream = new (new ByteBlockGenerator(yamlString));
+    // Parse the YAML byte block stream to a record type.
+    ServerConfig serverConfig3 = check yaml:parseStream(byteBlockStream);
+    io:println(serverConfig3);
+}
+
+// Defines a class called `ByteBlockGenerator`, which is stream implementer with `next()` method.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:33]:
```suggestion
// Defines a class called `ByteBlockGenerator`, which is stream implementor with a `next()` method.
```

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+// This will be invoked when the `next()` method of the stream gets invoked.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:34]:
```suggestion
// This `next()` method is called when iterating over a stream created with a `ByteBlockGenerator` value.
```

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+        host: "localhost"
+        port: 8080
+        remotePorts: [9000, 9001, 9002, 9003]
+        protocol: "http"
+        database:
+          dbName: "testdb"
+          username: "dbuser"
+          password: "dbpassword"`;
+
+    // Based on the expected type, it selectively converts the YAML string to the record type.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:28]:
```suggestion
    // Based on the expected type, this function selectively constructs the record type from the YAML string.
```

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+|};
+
+final string yamlString = string
+        `
+        host: "localhost"
+        port: 8080
+        protocol: "http"`;
+
+public function main() returns error? {
+    // Parse the YAML string to a record type.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:17]:
```suggestion
    // Parse the YAML string as a record.
```

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    ServerConfig serverConfig1 = check yaml:parseString(yamlString);
+    io:println(serverConfig1);
+
+    byte[] yamlByteArr = yamlString.toBytes();
+    // Parse the YAML byte array to a record type.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:22]:
```suggestion
    // Parse the YAML byte array as a record.
```

File: examples/yaml-to-anydata/yaml_to_anydata.bal [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+    ServerConfig serverConfig2 = check yaml:parseBytes(yamlByteArr);
+    io:println(serverConfig2);
+
+    stream<byte[], error?> byteBlockStream = new (new ByteBlockGenerator(yamlString));
+    // Parse the YAML byte block stream to a record type.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.bal:27]:
```suggestion
    // Parse the YAML byte block stream as a record.
```

File: examples/yaml-to-anydata/yaml_to_anydata.metatags [hunk 29]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates how to convert a YAML source, which can be provided as a string, byte array, or byte block stream, into a subtype of Ballerina anydata.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.metatags:1]:
```suggestion
description: This BBE demonstrates how to parse a YAML source, which can be provided as a string, byte array, or byte block stream, as a value that belongs to a subtype of Ballerina anydata.
```

File: examples/index.json [hunk 30]
```diff
@@ -4614,11 +4614,39 @@
         "name": "Record to XML conversion",
         "url": "xml-from-record-conversion",
         "verifyBuild": true,
-        "verifyOutput": true,
+        "verifyOutput": false,
```
Comment (MaryamZi) [inline examples/index.json:4617]: Why this change?

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+        host: "localhost"
+        port: 8080
+        remotePorts: [9000, 9001, 9002, 9003]
+        protocol: "http"
+        database:
+          dbName: "testdb"
+          username: "dbuser"
+          password: "dbpassword"`;
+
+    // Based on the expected type, this function selectively constructs the record type from the YAML string.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:28]:
```suggestion
    // Based on the expected type, this function selectively constructs the record from the YAML string.
```

or

```suggestion
    // Based on the expected record type, this function selectively constructs a value from the YAML string.
```

File: examples/yaml-to-anydata/yaml_to_anydata.metatags [hunk 32]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates how to parse a YAML source, which can be provided as a string, byte array, or byte block stream, as a value that belongs to a subtype of Ballerina anydata.
```
Comment (MaryamZi) [inline examples/yaml-to-anydata/yaml_to_anydata.metatags:1]:
```suggestion
description: This BBE demonstrates how to parse a YAML source, which can be provided as a string, byte array, or byte block stream, as a value that belongs to a subtype of anydata.
```

File: examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal (hunk 31 above)
Comment (MaryamZi) [reply examples/yaml-to-anydata-with-projection/yaml_to_anydata_with_projection.bal:28]: Reason is because we are creating a value of the specific type, not the type itself.

File: examples/index.json (hunk 30 above)
Comment (LakshanWeerasinghe) [reply examples/index.json:4617]: This test was failing due to its output cannot be verified. So decided to make it false.

Comment (MaryamZi) [reply examples/index.json:4617]: Why aren't we fixing output verification instead? Isn't the issue that the new xmldata version isn't updated in the distribution? Let's undo this change and send a separate PR for that (or get the library owner/RM to look into it) first.

**Review by MaryamZi — APPROVED** [review]
